File: frontend/src/lib/scan/holdoff.ts
/**
 * The payload hold-off — one mechanism, used at two windows.
 *
 * `PLAN.md` calls for a **3-second payload-hash hold-off** in the decoder so
 * that one label held in front of the camera does not fire five resolves, and
 * for duplicate scans **within ~2 s, or during an in-flight commit**, to be
 * dropped "by the same debounce as the decoder". This class is that debounce;
 * the two call sites differ only in their window and in whether a repeat sighting
 * pushes the window forward.
 *
 * Keyed per payload rather than on a single most-recent slot, because two labels
 * alternating in view must each be held off independently — a single slot would
 * let A, B, A, B fire four times.
 */

/** Decoder window: a label sitting in the frame resolves once. */
export const DECODER_HOLD_OFF_MS = 3_000;

/** Scan-session window: a deliberate re-scan is dropped if it is this quick. */
export const SCAN_DEBOUNCE_MS = 2_000;

export interface HoldOffOptions {
  /** Injected in tests; defaults to the wall clock. */
  readonly now?: () => number;
  /**
   * Whether a suppressed sighting restarts the window.
   *
   * `false` (default) is the literal reading of "duplicate scans within ~2 s are
   * dropped": the window runs from the sighting that was *admitted*.
   *
   * `true` is what the decoder needs. At ~10 frames a second a label held up for
   * fifteen seconds is sighted 150 times, and with a fixed window that still
   * fires five resolves — exactly what the hold-off exists to prevent. Pushing
   * the window forward on every sighting means the label fires once and cannot
   * fire again until it has been out of frame for the full window.
   */
  readonly refreshWhileSuppressed?: boolean;
}

export class PayloadHoldOff {
  readonly #windowMs: number;
  readonly #now: () => number;
  readonly #refresh: boolean;
  readonly #seen = new Map<string, number>();
  #blocked = false;

  constructor(windowMs: number, options: HoldOffOptions = {}) {
    this.#windowMs = windowMs;
    this.#now = options.now ?? (() => Date.now());
    this.#refresh = options.refreshWhileSuppressed ?? false;
  }

  /**
   * Whether this payload should be acted on now. Records the sighting.
   *
   * Returns `false` while blocked, and `false` for a payload seen inside the
   * window.
   */
  admit(payload: string): boolean {
    const at = this.#now();
    if (this.#blocked) {
      // A sighting during a commit is still a sighting: remember it so that
      // letting go of the block does not immediately fire the label the user is
      // still holding in front of the lens.
      if (this.#refresh) {
        this.#seen.set(payload, at);
      }
      return false;
    }

    this.#prune(at);
    const previous = this.#seen.get(payload);
    if (previous !== undefined && at - previous < this.#windowMs) {
      if (this.#refresh) {
        this.#seen.set(payload, at);
      }
      return false;
    }
    this.#seen.set(payload, at);
    return true;
  }

  /**
   * Suppress everything — used while a commit is in flight, so a double scan
   * cannot queue a second movement behind the first.
   */
  block(): void {
    this.#blocked = true;
  }

  unblock(): void {
    this.#blocked = false;
  }

  get blocked(): boolean {
    return this.#blocked;
  }

  /** Forget every sighting, e.g. when the camera is restarted. */
  forget(): void {
    this.#seen.clear();
  }

  #prune(at: number): void {
    for (const [payload, seenAt] of this.#seen) {
      if (at - seenAt >= this.#windowMs) {
        this.#seen.delete(payload);
      }
    }
  }
}

File: deviceagent/agent/holdoff.py
"""The 400 ms hold-off, keyed per payload. Python sibling of the PWA's copy.

`frontend/src/lib/scan/holdoff.ts` is the same mechanism with the same semantics,
and `frontend/src/lib/scan/feedback.ts` fixes the window this module defaults to.
The number is **not invented here**: PLAN.md's tag-provisioning walk specifies a
"400 ms debounce (so the same tag can't double-fire while still in range)" before
the cursor auto-advances, and the frontend already reuses it for decode feedback.
Reusing it a third time is the point — one number, one meaning, so a user who has
learned how fast the bench responds is not surprised by which screen they are on.

Two implementations rather than one because the two runtimes are genuinely
separate (a browser and a daemon on the Pi), and a shared constant would have to
travel through the API schema, which is for the data model and not for UI timing.
`tests/test_holdoff.py` pins the semantics to the TypeScript file's.

Keyed per payload rather than on one most-recent slot, for the same reason as the
TS version: two alternating keys must each be held off independently, or A, B, A,
B fires four times.
"""

from __future__ import annotations

import time
from collections.abc import Callable

#: PLAN.md's cabinet-binding recipe, and `FEEDBACK_DEBOUNCE_MS` in the frontend.
DEFAULT_WINDOW_MS = 400


class HoldOff:
    """Admit a key at most once per window.

    The window runs from the sighting that was **admitted**, not from the last one
    seen — the literal reading of "duplicate within the window is dropped", and the
    behaviour a Commit button needs: a double-tap is one commit, and a deliberate
    second identical action a second later is honoured rather than being held off
    for as long as the user keeps tapping.
    """

    def __init__(
        self,
        window_ms: int = DEFAULT_WINDOW_MS,
        *,
        clock: Callable[[], float] = time.monotonic,
    ) -> None:
        if window_ms < 0:
            raise ValueError("window_ms cannot be negative")
        self._window_s = window_ms / 1000.0
        # Monotonic, not wall clock: an NTP step or a DST change must not open or
        # close the window, and a station Pi does step its clock at boot.
        self._clock = clock
        self._admitted: dict[str, float] = {}

    def admit(self, key: str) -> bool:
        """Whether to act on this key now. Records the sighting when it says yes."""
        now = self._clock()
        last = self._admitted.get(key)
        if last is not None and now - last < self._window_s:
            return False
        self._admitted[key] = now
        return True

    def forget(self, key: str) -> None:
        """Drop a key's window, so the next sighting is admitted immediately.

        Called when the thing the key identified is gone — a session ending, an
        idempotency key rotating. Without it the map grows for the lifetime of the
        process, and at a bench that is weeks.
        """
        self._admitted.pop(key, None)

    def clear(self) -> None:
        self._admitted.clear()
